python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -926,9 +939,18 @@ def _get_nodes_needed_for_request_resources(
         # Since it is sorted based on last used, we "keep" nodes that are
         # most recently used when we binpack. We assume get_bin_pack_residual
         # is following the given order here.
-        used_resource_requests: List[ResourceDict]
+        used_resource_requests: List[ResourceDict] = copy.deepcopy(max_node_resources)
+
+        for bundles in strict_spreads:
+            unfulfilled, updated_node_resources = get_bin_pack_residual(
+                used_resource_requests, bundles, strict_spread=True
+            )
+            if unfulfilled:
+                continue
+            used_resource_requests = updated_node_resources
+
         _, used_resource_requests = get_bin_pack_residual(
-            max_node_resources, self.load_metrics.get_resource_requests()
+            used_resource_requests, resource_demands
         )
         # Remove the first entry (the head node).
         max_node_resources.pop(0)
```

Comment:
I would recommend that the following logic not rely on the assumption that `get_bin_pack_residual` returns a list in the same order as this function. If it increases a lot of complexity to avoid the assumption, we at least add comments and unit tests for `get_bin_pack_residual` to ensure it returns a list of nodes with the same order.

Reply:
A test is added.